libc/src/__support/OSUtil/linux/random.cpp

```diff
@@ -0,0 +1,359 @@
+    cpp::lock_guard guard{mutex};
+    // there should be no need to check this pushing
+    freelist.push_unchecked(state);
+  }
+  ~RandomStateFactory() {
+    for (auto *allocation : allocations)
+      munmap(allocation, page_size * pages_per_allocation);
+  }
+
+public:
+  static void *acquire_global() {
+    auto *factory = instance();
+    if (!factory)
+      return nullptr;
+    return factory->acquire();
+  }
+  static void release_global(void *state) {
+    auto *factory = instance();
+    if (!factory)
+      return;
+    factory->release(state);
+  }
+  static size_t size_of_opaque_state() {
+    return instance()->params.size_of_opaque_state;
+  }
+  static void postfork_cleanup();
+};
+
+thread_local bool fork_inflight = false;
+thread_local void *tls_state = nullptr;
+alignas(RandomStateFactory) static char factory_storage[sizeof(
+    RandomStateFactory)]{};
+static CallOnceFlag factory_onceflag = callonce_impl::NOT_CALLED;
+static bool factory_valid = false;
+
+RandomStateFactory *RandomStateFactory::instance() {
+  callonce(&factory_onceflag, []() {
+    auto *factory = new (factory_storage) RandomStateFactory();
+    factory_valid = factory->prepare();
+    if (factory_valid)
+      atexit([]() {
+        auto factory = reinterpret_cast<RandomStateFactory *>(factory_storage);
+        factory->~RandomStateFactory();
+        factory_valid = false;
+      });
+  });
+  return factory_valid ? reinterpret_cast<RandomStateFactory *>(factory_storage)
+                       : nullptr;
+}
+
+void RandomStateFactory::postfork_cleanup() {
+  if (factory_valid)
+    reinterpret_cast<RandomStateFactory *>(factory_storage)
+        ->~RandomStateFactory();
+  factory_onceflag = callonce_impl::NOT_CALLED;
+  factory_valid = false;
+}
+
+void *acquire_tls() {
+  if (fork_inflight)
+    return nullptr;
+  // previous acquire failed, do not try again
+  if (tls_state == MAP_FAILED)
+    return nullptr;
+  // first acquirement
+  if (tls_state == nullptr) {
+    tls_state = RandomStateFactory::acquire_global();
+    // if still fails, remember the failure
+    if (tls_state == nullptr) {
+      tls_state = MAP_FAILED;
+      return nullptr;
+    } else {
+      // register the release callback.
+      if (__cxa_thread_atexit_impl(
+              [](void *s) { RandomStateFactory::release_global(s); }, tls_state,
+              __dso_handle)) {
+        RandomStateFactory::release_global(tls_state);
+        tls_state = MAP_FAILED;
+        return nullptr;
+      }
+    }
+  }
+  return tls_state;
+}
+
+template <class F> void random_fill_impl(F gen, void *buf, size_t size) {
+  auto *buffer = reinterpret_cast<uint8_t *>(buf);
+  while (size > 0) {
+    ssize_t len = gen(buffer, size);
+    if (len == -1) {
+      if (libc_errno == EINTR)
+        continue;
+      break;
+    }
+    size -= len;
+    buffer += len;
+  }
+}
+} // namespace
+
+void random_fill(void *buf, size_t size) {
+  ErrnoProtect protect;
+  void *state = acquire_tls();
```

Comment:
random_fill should be async-signal-safe. That means, it must be safe to call random_fill from a signal handler, even when random_fill is already executing on the same thread.

This code does not seem to currently consider this yet. Here, it's not tracking whether the random state is in-use, so a signal handler re-entrantly calling random_fill might reuse the same state. It should fallback to a syscall if the state is in-use.